File: apps/desktop/src/features/workbench/review/application/ports.ts
// @author kongweiguang
// SPDX-License-Identifier: GPL-3.0-or-later

import type {
  ReviewAction,
  ReviewApplyResult,
  ReviewCancelResult,
  ReviewCatalog,
  ReviewFileDiff,
  ReviewInvalidatedEvent,
  ReviewSnapshot,
  ReviewSource,
  ReviewTarget,
} from "../domain/types";

interface ReviewCatalogInput {
  workspaceId: string;
  maxCommits?: number;
}

interface ReviewSnapshotInput {
  workspaceId: string;
  source: ReviewSource;
}

interface ReviewFileDiffInput {
  workspaceId: string;
  source: ReviewSource;
  revision: string;
  fileId: string;
}

interface ReviewApplyInput {
  workspaceId: string;
  source: ReviewSource;
  revision: string;
  action: ReviewAction;
  target: ReviewTarget;
  operationId: string;
}

interface ReviewCancelInput {
  workspaceId: string;
  operationId: string;
}

type ReviewUnsubscribe = () => void | Promise<void>;

/** application 只依赖该窄端口，Tauri adapter 由 App 组合并注入。 */
export interface ReviewPort {
  catalog(input: ReviewCatalogInput): Promise<ReviewCatalog>;
  snapshot(input: ReviewSnapshotInput): Promise<ReviewSnapshot>;
  fileDiff(input: ReviewFileDiffInput): Promise<ReviewFileDiff>;
  apply(input: ReviewApplyInput): Promise<ReviewApplyResult>;
  cancel(input: ReviewCancelInput): Promise<ReviewCancelResult>;
  subscribeInvalidated(
    listener: (event: ReviewInvalidatedEvent) => void,
  ): Promise<ReviewUnsubscribe>;
}

export interface ReviewFailure {
  code: string;
  message: string;
  retryable: boolean;
}

const REVIEW_FAILURES: Readonly<Record<string, Omit<ReviewFailure, "code">>> = {
  REVIEW_STALE: { message: "工作区已经变化，请重新读取 Review。", retryable: true },
  REVIEW_CONFLICT: { message: "当前文件存在重叠修改，无法安全应用。", retryable: false },
  REVIEW_READ_ONLY: { message: "此 Review 来源为只读。", retryable: false },
  REVIEW_UNAVAILABLE: { message: "当前轮次没有可精确归因的变更。", retryable: false },
  REVIEW_LIMIT: { message: "Review 内容超过安全上限。", retryable: false },
  NOT_GIT_REPOSITORY: { message: "当前目录不是 Git 工作区，审查不可用。", retryable: false },
  INVALID_INPUT: { message: "Review 请求参数无效。", retryable: false },
  WORKSPACE_ESCAPE: { message: "请求路径不在工作区内。", retryable: false },
  HOST_BUSY: { message: "工作区正被其他操作占用，请稍后重试。", retryable: true },
  CANCELLED: { message: "Review 操作已取消。", retryable: true },
  RUNTIME_UNAVAILABLE: { message: "运行时暂不可用，请稍后重试。", retryable: true },
  RUNTIME_PROTOCOL_ERROR: { message: "运行时返回的数据无效。", retryable: false },
};

/** 只读取稳定错误码并生成本地安全文案，禁止 native 诊断进入 React state。 */
export function normalizeReviewFailure(error: unknown): ReviewFailure {
  const candidate =
    error !== null && typeof error === "object" ? (error as { code?: unknown }) : undefined;
  const code = typeof candidate?.code === "string" ? candidate.code : "RUNTIME_UNAVAILABLE";
  const safe = REVIEW_FAILURES[code] ?? REVIEW_FAILURES["RUNTIME_UNAVAILABLE"]!;
  return { code, ...safe };
}

/** 协议归属校验失败时使用固定错误，避免把 native DTO 内容拼入 UI。 */
export function reviewProtocolFailure(message = "运行时返回的数据无效。"): ReviewFailure {
  return { code: "RUNTIME_PROTOCOL_ERROR", message, retryable: false };
}

/** 为缺少 App 注入的场景提供 fail-closed 端口，不在 feature 内偷偷创建 Tauri adapter。 */
function unavailable(): Promise<never> {
  return Promise.reject({ code: "RUNTIME_UNAVAILABLE" });
}

export const unavailableReviewPort: ReviewPort = {
  catalog: unavailable,
  snapshot: unavailable,
  fileDiff: unavailable,
  apply: unavailable,
  cancel: unavailable,
  subscribeInvalidated: unavailable,
};

File: src-tauri/src/review/application/ports.rs
// SPDX-License-Identifier: GPL-3.0-or-later
// @author kongweiguang

// Review 应用层端口。

use super::ReviewError;
use crate::review::domain::{
    ReviewAction, ReviewCatalog, ReviewCatalogLimit, ReviewFile, ReviewRevision, ReviewSnapshot,
    ReviewSource, ReviewTarget,
};
use tokio_util::sync::CancellationToken;

/// application 对原生 Review 能力的唯一端口。
///
/// 端口刻意保留原子的 `apply_transaction`，因为路径锁、临时 index 与 worktree
/// recovery 必须在同一 native 临界区完成；逐个暴露句柄会破坏锁生命周期。
pub(crate) trait ReviewNativePort: Send + Sync {
    /// 读取有界 repository catalog，不允许调用方传入任意 Git 参数。
    fn catalog(
        &self,
        max_commits: ReviewCatalogLimit,
        cancellation: &CancellationToken,
    ) -> Result<ReviewCatalog, ReviewError>;

    /// 物化一个 source 的权威快照；revision 必须覆盖所有可变选择证据。
    fn snapshot(
        &self,
        source: &ReviewSource,
        cancellation: &CancellationToken,
    ) -> Result<ReviewSnapshot, ReviewError>;

    /// 在应用层已根据 snapshot 解析 file id 后，按需补充 metadata-only diff。
    fn materialize_file(
        &self,
        source: &ReviewSource,
        file: &ReviewFile,
        cancellation: &CancellationToken,
    ) -> Result<ReviewFile, ReviewError>;

    /// 在路径锁内重新检查 expected revision，并原子执行 Git/index/worktree mutation。
    ///
    /// `initial` 是 application 已验收的同一快照，用于确定锁集合；实现仍必须在锁内
    /// 重新物化并做 CAS。失败时必须回滚 worktree 并丢弃临时 index。
    fn apply_transaction(
        &self,
        source: &ReviewSource,
        expected_revision: &ReviewRevision,
        action: ReviewAction,
        target: &ReviewTarget,
        initial: &ReviewSnapshot,
        cancellation: &CancellationToken,
    ) -> Result<ReviewSnapshot, ReviewError>;
}
